core/src/main/java/org/jdbi/v3/core/mapper/OptionalMapperFactory.java

```diff
@@ -38,28 +42,59 @@
  * </ul>
  */
 class OptionalMapperFactory implements ColumnMapperFactory {
-    private final Map<Class<?>, ColumnMapper<?>> mappers = new IdentityHashMap<>();
-
-    OptionalMapperFactory() {
-        mappers.put(OptionalInt.class, optionalMapper(ResultSet::getInt, OptionalInt::empty, OptionalInt::of));
-        mappers.put(OptionalLong.class, optionalMapper(ResultSet::getLong, OptionalLong::empty, OptionalLong::of));
-        mappers.put(OptionalDouble.class, optionalMapper(ResultSet::getDouble, OptionalDouble::empty, OptionalDouble::of));
-    }
+    private static final Map<Class<?>, OptionalStrategy> STRATEGIES =
+            Collections.unmodifiableMap(Arrays.stream(OptionalStrategy.values())
+                .collect(Collectors.toMap(
+                        OptionalStrategy::optClass,
+                        Function.identity(),
+                        (u, v) -> {
+                            throw new UnsupportedOperationException();
+                        },
+                        IdentityHashMap::new)));
 
     @Override
     public Optional<ColumnMapper<?>> build(Type type, ConfigRegistry config) {
-        Class<?> rawType = getErasedType(type);
+        return Optional.ofNullable(STRATEGIES.get(getErasedType(type)))
+                .map(strategy -> strategy.create(type, config));
+    }
+
+    private enum OptionalStrategy {
+        OBJ(Optional.class) {
+            @Override
+            ColumnMapper<?> create(Type type, ConfigRegistry config) {
+                final ColumnMapper<?> mapper = config.get(ColumnMappers.class).findFor(
+                        GenericTypes.findGenericParameter(type, Optional.class)
+                            .orElseThrow(() -> new NoSuchMapperException("No mapper for raw Optional type")))
+                        .orElseThrow(() -> new NoSuchMapperException("No mapper for type " + type + " nested in Optional"));
+                return (r, i, ctx) -> (Optional<?>) Optional.ofNullable(mapper.map(r, i, ctx));
+            }
+        },
+        INT(OptionalInt.class, optionalMapper(ResultSet::getInt, OptionalInt::empty, OptionalInt::of)),
+        LNG(OptionalLong.class, optionalMapper(ResultSet::getLong, OptionalLong::empty, OptionalLong::of)),
+        DBL(OptionalDouble.class, optionalMapper(ResultSet::getDouble, OptionalDouble::empty, OptionalDouble::of));
+
+        final Class<?> optClass;
+        final ColumnMapper<?> impl;
 
-        if (rawType == Optional.class) {
-            return Optional.of(OptionalMapper.of(type));
+        OptionalStrategy(Class<?> optClass) {
+            this(optClass, null);
+        }
+        OptionalStrategy(Class<?> optClass, ColumnMapper<?> impl) {
```

Comment:
call me a twit, but I would like to see the conventional spacing between methods respected